src/t8_forest/t8_forest_vtk.cxx

```diff
@@ -147,45 +158,53 @@ t8_write_vtk_via_API (t8_forest_t forest, const char *fileprefix)
   vtkNew < vtkWedge > prism;
   vtkNew < vtkTetra > tet;
 
-  std::string filename = fileprefix;
-  // Append process number
-  filename += std::to_string(forest->mpirank);
-
-  filename += ".vtu";
   cmesh = t8_forest_get_cmesh (forest);
-  int         *cellTypes =
-    T8_ALLOC (int, t8_forest_get_num_element(forest));
+  /* 
+   * The cellTypes Array stores the element types as integers(see vtk doc).
+   * We have to Allocate memory since we do not know the size the array
+   * needs to be before running the code to evaluate the number of elements
+   * in the forest. We have to free the memory when we are done. 
+   */
+  int                *cellTypes =
+    T8_ALLOC (int, t8_forest_get_num_element (forest));
+
+/* We iterate over all local trees*/
+  for (itree = 0; itree < t8_forest_get_num_local_trees (forest); itree++) {
+/* 
+ * We get the vertices, the current tree, the scheme for this tree
+ * and the number of elements in this tree. We need the vertices of
+ * the tree to get the coordinates of the elements later. We need
+ * the number of elements in this to iterate over all of them.  
+ */
+    vertices = t8_forest_get_tree_vertices (forest, itree);
+    t8_eclass_scheme_c *scheme =
+      t8_forest_get_eclass_scheme (forest, t8_forest_get_tree_class (forest,
+                                                                     itree));
+    t8_locidx_t         elems_in_tree =
+      t8_forest_get_tree_num_elements (forest, itree);
 
-  for (itree = 0;
-       (int) itree < t8_forest_get_num_local_trees (forest); itree++) {
+    /* We iterate over all elements in the tree */
+    for (ielement = 0; ielement < elems_in_tree; ielement++) {
 
-    vertices = t8_forest_get_tree_vertices(forest, itree);
-    tree = t8_forest_get_tree (forest, itree);
-    t8_eclass_scheme_c *scheme =
-      t8_forest_get_eclass_scheme (forest, tree->eclass);
-    t8_locidx_t elems_in_tree =
-      (t8_locidx_t) t8_element_array_get_count (&tree->elements);
-    for (ielement = 0;
-         ielement < elems_in_tree;
-         ielement++) {
       t8_element_t       *element =
         t8_forest_get_element_in_tree (forest, itree, ielement);
-      T8_ASSERT(element!=NULL);
+      T8_ASSERT (element != NULL);
       t8_element_shape_t  element_shape = scheme->t8_element_shape (element);
 
-      for (ivertex = 0; ivertex < t8_eclass_num_vertices[tree->eclass];
+      /* For each element we iterate over all points */
+      for (ivertex = 0; ivertex < scheme->t8_element_num_corners (element);
```

Comment:
As a tip:

You should avoid using a function evaluation in a for loop.
The compiler does not know that the value does not change and thus the function is called in every step of the loop.
It's way better for the performance to compute the function once, store the value in a variable and then use the varible in the for-loop.